NNDL ML HONORS ASSIGNMENT <br>
Irene Elizabeth Sabu<br>
TVE19CS038<br>
S5 CSE<br>
Access the Github Repo From : [This Github Repo]()

In [32]:
from numpy import mean,std
from matplotlib import pyplot as plt
from sklearn.model_selection import KFold
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D,MaxPooling2D,Dense,Flatten
from tensorflow.keras.optimizers import Adam

In [33]:
# load dataset
(trainX, trainy), (testX, testy) = mnist.load_data()
# summarize loaded dataset
print('Train: X=%s, y=%s' % (trainX.shape, trainy.shape))
print('Test: X=%s, y=%s' % (testX.shape, testy.shape))

Train: X=(60000, 28, 28), y=(60000,)
Test: X=(10000, 28, 28), y=(10000,)


In [34]:
(trainX, trainY), (testX, testY) = mnist.load_data()

# reshape dataset to have a single channel
trainX = trainX.reshape((trainX.shape[0], 28, 28, 1))
testX = testX.reshape((testX.shape[0], 28, 28, 1))

In [35]:
from tensorflow.keras.utils import to_categorical

# one hot encode target values
trainY = to_categorical(trainY)
testY = to_categorical(testY)

In [36]:
# convert from integers to floats
train_norm = trainX.astype('float32')
test_norm = testX.astype('float32')
# normalize to range 0-1
trainX = train_norm / 255.0
testX = test_norm / 255.0

**MODEL 1**

In [45]:
def define_model_1():
	model = Sequential()
	model.add(Flatten(input_shape=(28, 28)))
	model.add(Dense(200, activation='relu'))
	model.add(Dense(10, activation='softmax'))
	# compile model
	opt = Adam(learning_rate=0.01)
	model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
	return model

In [46]:
# evaluate a model using k-fold cross-validation
def evaluate_model_1():
    scores, histories = list(), list()
    # define model
    model = define_model_1()
    print(model.summary())
    # fit model
    history = model.fit(trainX, trainY, epochs=10, batch_size=32, validation_data=(testX, testY), verbose=1)
    # evaluate model
    _, acc = model.evaluate(testX, testY, verbose=1)
    print('> %.3f' % (acc * 100.0))
    # stores scores
    scores.append(acc)
    histories.append(history)
    return scores, histories

In [47]:
scores, histories = evaluate_model_1()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_8 (Flatten)         (None, 784)               0         
                                                                 
 dense_18 (Dense)            (None, 200)               157000    
                                                                 
 dense_19 (Dense)            (None, 10)                2010      
                                                                 
Total params: 159,010
Trainable params: 159,010
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2458 - accuracy: 0.9283 - val_loss: 0.1889 - val_accuracy: 0.9467
Epoch 2/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1625 - accuracy: 0.9564 - val_loss: 0.1905 - val_accuracy: 0.9519
Epoch 3/10
1875/187

**MODEL 2**

In [79]:
def define_model_2():
    model = Sequential()
    model.add(Flatten(input_shape=(28, 28)))
    model.add(Dense(220, activation='relu'))
    model.add(Dense(150, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(10, activation='softmax'))
    # compile model
    opt = Adam(learning_rate=0.01)
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [80]:
# evaluate a model using k-fold cross-validation
def evaluate_model_2():
    scores, histories = list(), list()
    # define model
    model = define_model_2()
    print(model.summary())
    # fit model
    history = model.fit(trainX, trainY, epochs=10, batch_size=32, validation_data=(testX, testY), verbose=1)
    # evaluate model
    _, acc = model.evaluate(testX, testY, verbose=1)
    print('> %.3f' % (acc * 100.0))
    # stores scores
    scores.append(acc)
    histories.append(history)
    return scores, histories

Execute the Second Model Training

In [81]:
scores, histories = evaluate_model_2()

Model: "sequential_18"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_18 (Flatten)        (None, 784)               0         
                                                                 
 dense_56 (Dense)            (None, 220)               172700    
                                                                 
 dense_57 (Dense)            (None, 150)               33150     
                                                                 
 dense_58 (Dense)            (None, 64)                9664      
                                                                 
 dense_59 (Dense)            (None, 10)                650       
                                                                 
Total params: 216,164
Trainable params: 216,164
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/10
1875/1875 [====================

# Report

Two Deep Learning Models with softmax layer activation at the top and 'relu' on other layers using Neural Network architecture was created on the  MNIST dataset.  

MODEL 1:
<br>Number of NN layers = 2
<br>Validation Accuracy = 96.590
<br>Validation Loss = 0.1986
<br>Number of Trainable Parameters = 159,010

MODEL 2:
<br>Number of NN layers = 4
<br>Validation Accuracy = 96.950
<br>Validation Loss =  0.1613
<br>Number of Trainable Parameters = 216,164

Model 2 has better validation accuracy than Model 1 hence Model 2 has better performance.